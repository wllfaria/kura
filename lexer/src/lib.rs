use miette::{Error, LabeledSpan};

mod token;

use token::{NumSize, Token};

struct Lexer<'lex> {
    pos: usize,
    source: &'lex str,
    complete_source: &'lex str,
}

impl<'lex> Lexer<'lex> {
    pub fn new(source: &'lex str) -> Self {
        Self {
            pos: 0,
            source,
            complete_source: source,
        }
    }
}

impl<'lex> Lexer<'lex> {
    fn lex_numerals(&mut self) -> Result<Token<'lex>, Error> {
        let end_of_numeral = self
            .source
            .find(|c| !matches!(c, '-' | '_' | '.' | '0'..='9'))
            .unwrap_or(self.source.len());

        let mut literal = &self.source[..end_of_numeral];

        let mut dotted = literal.splitn(3, '.');
        match (dotted.next(), dotted.next(), dotted.next()) {
            (Some(one), Some(two), Some(_)) => {
                // numeric literal contains two or more dots, so we only lex
                // from the first numerical to the last numerical before the
                // second dot, <one.len()>.<two.len()>
                //
                // 123.456.789 -> 123.456 (leave .789) to be lexed after
                literal = &literal[..one.len() + 1 + two.len()];
            }
            _ => {
                // literal is either a integer, or a valid floating pointer
                // numeral, so we don't have to do anything here
            }
        }

        let mut dashed = literal.splitn(3, '-');
        match (dashed.next(), dashed.next(), dashed.next()) {
            (Some(""), Some(two), Some(_)) => {
                // numeric literal contain two or more dashes, but one of the
                // dashes is the first character, meaning we should only lex the
                // middle part as a signed integer
                //
                // -123.456-123 ("", "123.456", "123") (result of splitn)
                literal = &literal[..two.len() + 1]; // +1 to account for the `-`
            }
            (Some(one), Some(_), Some(_)) => {
                // numeric literal has two or more dashes, and don't start with a
                // dash, so we only parse the first part and ignore the rest.
                //
                // 123-456.78-9 ("123", "456.78", "9")
                literal = &literal[..one.len()];
            }
            _ => {
                // numeral is a unsigned integer, so we don't have to do
                // anything here
            }
        }

        // NOTE: we have to update the lexer state here before we potentially
        // change the size of the literal after trimming underscores
        let bytes_eaten = literal.len();
        self.pos += bytes_eaten;
        self.source = &self.source[bytes_eaten..];

        // we accept underscores in numerals, but rust parser don't, so we
        // trim here before parsing to rust's parser
        let literal = literal.replace('_', "");

        // numbers are allowed to potentially have a postfix specifying its size
        // like `1234usize`
        let end_of_postfix = self
            .source
            .find(|c: char| c.is_whitespace())
            .unwrap_or(self.source.len());

        let mut postfix = &self.source[..end_of_postfix];
        if postfix.ends_with(';') {
            postfix = postfix.trim_matches(';');
        }

        let size = match NumSize::try_from(postfix) {
            Ok(size) => Some(size),
            Err(_) => None,
        };

        if size.is_some() {
            self.source = &self.source[postfix.len()..];
            self.pos += postfix.len();
        }

        let is_signed = literal.contains('-');
        let is_float = literal.contains('.');
        match (is_float, is_signed) {
            (false, true) => Ok(Token::SignedInteger {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(e) => return Err(miette::miette! {
                        labels = vec![
                            LabeledSpan::at(self.pos - bytes_eaten..self.pos, "this numeric literal"),
                        ],
                        "{e}",
                    }.with_source_code(self.complete_source.to_string())),
                },
                size,
            }),
            (false, false) => Ok(Token::Integer {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(e) => return Err(miette::miette! {
                        labels = vec![
                            LabeledSpan::at(self.pos - bytes_eaten..self.pos, "this numeric literal"),
                        ],
                        "{e}",
                    }.with_source_code(self.complete_source.to_string())),
                },
                size,
            }),
            (true, _) => Ok(Token::Float {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(e) => return Err(miette::miette! {
                        labels = vec![
                            LabeledSpan::at(self.pos - bytes_eaten..self.pos, "this numeric literal"),
                        ],
                        "{e}",
                    }.with_source_code(self.complete_source.to_string())),
                },
                size,
            }),
        }
    }

    fn advance_by(&mut self, amount: usize) {
        self.source = &self.source[amount..];
        self.pos += amount;
    }
}

impl<'lex> Iterator for Lexer<'lex> {
    type Item = Result<Token<'lex>, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let mut chars = self.source.chars().peekable();
            let c = chars.next()?;

            let mut simple_token = |token: Token<'lex>| {
                self.advance_by(c.len_utf8());
                Some(Ok(token))
            };

            let mut token_from_pairs = |pairs: Vec<(char, Token<'lex>)>, default: Token<'lex>| {
                let Some(next) = chars.peek() else {
                    return (default, 1);
                };

                for pair in pairs {
                    if next == &pair.0 {
                        return (pair.1, 2);
                    }
                }

                (default, 1)
            };

            match c {
                c if c.is_whitespace() => {
                    self.advance_by(c.len_utf8());
                    continue;
                }
                c if c.is_numeric() => return Some(self.lex_numerals()),
                '(' => return simple_token(Token::LeftParen),
                ')' => return simple_token(Token::RightParen),
                '[' => return simple_token(Token::LeftBracket),
                ']' => return simple_token(Token::RightBracket),
                '{' => return simple_token(Token::LeftBrace),
                '}' => return simple_token(Token::RightBrace),
                ':' => return simple_token(Token::Colon),
                ';' => return simple_token(Token::SemiColon),
                ',' => return simple_token(Token::Comma),
                '.' => return simple_token(Token::Dot),
                '+' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::AddAssign)], Token::Plus);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '=' => {
                    let (token, advance) = token_from_pairs(
                        vec![('=', Token::EqualEqual), ('>', Token::ThickArrow)],
                        Token::Equal,
                    );
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '*' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::MultiplyAssign)], Token::Star);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '&' => return simple_token(Token::Ampersand),
                '/' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::DivideAssign)], Token::Slash);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '!' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::NotEqual)], Token::Bang);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '<' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::LessEqual)], Token::Less);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '>' => {
                    let (token, advance) =
                        token_from_pairs(vec![('=', Token::GreaterEqual)], Token::Greater);
                    self.advance_by(advance);
                    return Some(Ok(token));
                }
                '-' => match chars.peek() {
                    Some(c) if c.is_numeric() => return Some(self.lex_numerals()),
                    Some('=') => {
                        self.source = &self.source[c.len_utf8() + 1..];
                        self.pos += 2;
                        return Some(Ok(Token::MinusAssign));
                    }
                    _ => return simple_token(Token::Minus),
                },
                _ => {
                    let next_whitespace = self
                        .source
                        .find(|c| !matches!(c, 'a'..='z' | 'A'..='Z' | '_' | '0'..='9'))
                        .unwrap_or(self.source.len());

                    let identifier = &self.source[..next_whitespace];

                    self.advance_by(next_whitespace);

                    return Some(Ok(Token::identifier_from(identifier)));
                }
            };
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_sut(source: &str) -> Lexer<'_> {
        Lexer::new(source)
    }

    #[test]
    fn lexing_numericals() {
        let source = [
            // positive integers and floats
            "12346.67890",
            "1234567890",
            "9__876___5.43____210",
            "9_876_5.43_210",
            "9_____876__543____210",
            "9_876_543_210",
            // negative integers and floats
            "-9_____876__5.43____210",
            "-123_46.678_90",
            "-12346.67890",
            "-1234567890",
            "-9_____876__543____210",
            "-9_876_543_210",
            "-9_876_-5-43_210",
            // integers and floats with postfixes
            "-123456789i32",
            "123456789usize",
            "123.456789f64",
            "123.456789f64",
            "-12f8",
            "3.14159265358979323846264338327950288_f32",
            "3.141592653f32;",
        ];
        let source = source.join("\n");

        let mut numerals = vec![];

        for token in make_sut(&source) {
            numerals.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(numerals);
    }

    #[test]
    fn lexing_punctuations() {
        let source = [
            "()", "[]", "{}", ",", ".", "+", "-", "=", "*", "&", "*=", "+=", "-=", "/=", "!", "!=",
            "==", "<=", ">=", "<", ">", "/", ":", ";",
        ];
        let source = source.join(" ");

        let mut punctuations = vec![];
        for token in make_sut(&source) {
            punctuations.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(punctuations);
    }

    #[test]
    fn lexing_builtin_identifiers() {
        let source = [
            "var", "const", "match", "if", "else", "fun", "struct", "enum", "return",
        ];

        let source = source.join(" ");

        let mut builtin_identifiers = vec![];
        for token in make_sut(&source) {
            builtin_identifiers.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(builtin_identifiers);
    }

    #[test]
    fn lexing_simple_codeblock() {
        let source = [
            "fun calculate_circumference(diameter: f64) => f64 {",
            "   const pi = 3.14159265358979323846264338327950288_f32;",
            "   const radius = diameter / 2.0;",
            "   const circumference = 2.0 * pi * radius;",
            "",
            "   return circumference;",
            "}",
        ];

        let source = source.join("\n");

        let mut calculate_circumference_function = vec![];
        for token in make_sut(&source) {
            calculate_circumference_function.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(calculate_circumference_function);
    }
}
