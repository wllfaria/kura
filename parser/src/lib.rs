use lexer::{FloatSizes, IntSizes, Kind, Lexer, Operator, Primitive, Token, UIntSizes, Value};

use miette::{Error, LabeledSpan};

#[derive(Debug)]
pub enum BinaryOperator {
    Add,
    Sub,
    Mul,
    Div,
}

impl TryFrom<&Token<'_>> for BinaryOperator {
    type Error = Error;

    fn try_from(token: &Token) -> Result<Self, Self::Error> {
        match token.kind {
            Kind::Op(Operator::Plus) => Ok(BinaryOperator::Add),
            Kind::Op(Operator::Minus) => Ok(BinaryOperator::Sub),
            Kind::Op(Operator::Star) => Ok(BinaryOperator::Mul),
            Kind::Op(Operator::Slash) => Ok(BinaryOperator::Div),
            _ => miette::bail!("invalid binary operator"),
        }
    }
}

#[derive(Debug)]
pub enum Expression<'ast> {
    VarNode {
        name: &'ast str,
        value: Box<Expression<'ast>>,
    },
    UintLiteral {
        value: u64,
        size: Option<UIntSizes>,
    },
    FloatLiteral {
        value: f64,
        size: Option<FloatSizes>,
    },
    IntLiteral {
        value: i64,
        size: Option<IntSizes>,
    },
    BinaryOp {
        operator: BinaryOperator,
        lhs: Box<Expression<'ast>>,
        rhs: Box<Expression<'ast>>,
    },
}

pub struct Parser<'par> {
    source: &'par str,
    lexer: Lexer<'par>,
}

impl<'par> Parser<'par> {
    pub fn new(input: &'par str) -> Self {
        Self {
            source: input,
            lexer: Lexer::new(input),
        }
    }

    pub fn parse(mut self) -> Result<Vec<Expression<'par>>, Error> {
        let mut statements = vec![];

        loop {
            if self.lexer.is_empty() {
                break;
            }

            let statement = self.parse_statement()?;
            statements.push(statement);
        }

        Ok(statements)
    }

    pub fn parse_expression(&mut self) -> Result<Expression<'par>, Error> {
        self.parse_expression_with_precedence(0)
    }

    pub fn peek_for_binary_operator(&mut self) -> Result<Option<&Token<'par>>, Error> {
        match self.lexer.peek() {
            Some(result) => match result {
                Ok(token) if token.kind.is_binary_op() => Ok(Some(token)),
                Ok(_) => Ok(None),
                Err(_) => Ok(None),
            },

            None => Ok(None),
        }
    }

    pub fn parse_expression_with_precedence(
        &mut self,
        min_precedence: u8,
    ) -> Result<Expression<'par>, Error> {
        let mut lhs = self.parse_base_expression()?;

        while let Some(op) = self.peek_for_binary_operator()? {
            let (lhs_precedence, rhs_precedence) = op.kind.infix_precedence()?;
            if lhs_precedence < min_precedence {
                break;
            }

            let op = op.try_into()?;
            self.lexer.next().transpose()?;

            let rhs = self.parse_expression_with_precedence(rhs_precedence)?;

            lhs = Expression::BinaryOp {
                operator: op,
                lhs: Box::new(lhs),
                rhs: Box::new(rhs),
            };
        }

        Ok(lhs)
    }

    pub fn parse_base_expression(&mut self) -> Result<Expression<'par>, Error> {
        match self.lexer.next().transpose()? {
            Some(Token {
                kind: Kind::Value(Value::Primitive(Primitive::Float { value, .. })),
                ..
            }) => Ok(Expression::FloatLiteral { value, size: None }),
            Some(Token {
                kind: Kind::Value(Value::Primitive(Primitive::Int { value, .. })),
                ..
            }) => Ok(Expression::IntLiteral { value, size: None }),
            Some(Token {
                kind: Kind::Value(Value::Primitive(Primitive::UInt { value, .. })),
                ..
            }) => Ok(Expression::UintLiteral { value, size: None }),
            _ => todo!(),
        }
    }

    pub fn parse_statement(&mut self) -> Result<Expression<'par>, Error> {
        match self.lexer.next().transpose()? {
            Some(Token {
                kind: Kind::Value(Value::Primitive(Primitive::Float { value, .. })),
                ..
            }) => Ok(Expression::FloatLiteral { value, size: None }),
            Some(Token {
                kind: Kind::Value(Value::Primitive(Primitive::Int { value, .. })),
                ..
            }) => Ok(Expression::IntLiteral { value, size: None }),
            Some(Token {
                kind: Kind::Value(Value::Primitive(Primitive::UInt { value, .. })),
                ..
            }) => Ok(Expression::UintLiteral { value, size: None }),
            Some(Token {
                kind: Kind::Var, ..
            }) => self.parse_variable(),
            t => todo!("{t:?}"),
        }
    }

    pub fn parse_variable(&mut self) -> Result<Expression<'par>, Error> {
        let Some(name) = self.lexer.next().transpose()? else {
            return Err(
                miette::miette!("temporary error").with_source_code(self.source.to_string())
            );
        };

        let name = match name {
            Token {
                kind: Kind::Value(Value::Ident(name)),
                ..
            } => name,
            _ => panic!("invalid token where name of var should be"),
        };

        let Some(assign) = self.lexer.next().transpose()? else {
            return Err(
                miette::miette!("temporary error").with_source_code(self.source.to_string())
            );
        };

        if !matches!(
            assign,
            Token {
                kind: Kind::Op(Operator::Equal),
                ..
            }
        ) {
            return Err(
                miette::miette!("temporary error").with_source_code(self.source.to_string())
            );
        }

        let expression = Expression::VarNode {
            name,
            value: Box::new(self.parse_expression()?),
        };

        if self.lexer.next().transpose()?.is_some_and(|token| {
            !matches!(
                token,
                Token {
                    kind: Kind::Op(Operator::SemiColon),
                    ..
                }
            )
        }) {
            return Err(miette::miette! {
                labels = vec![
                    LabeledSpan::at(0..3, "this expression"),
                ],
                help = "you might have forgotten a `;` semicolon",
                "Unterminated expression",
            }
            .with_source_code(self.source.to_string()));
        }

        Ok(expression)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_sut(source: &str) -> Parser<'_> {
        Parser::new(source)
    }

    #[test]
    fn tests() {
        let source = [
            "var some_name = 1 + 2 * 3 - 4;",
            "var another_name = 123.4;",
        ];
        let source = source.join("\n");

        let result = make_sut(&source).parse().unwrap();

        insta::assert_debug_snapshot!(result);
    }
}
