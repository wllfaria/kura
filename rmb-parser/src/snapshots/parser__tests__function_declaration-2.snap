---
source: parser/src/lib.rs
expression: fun_expr
---
fun calculate_circumference(diameter: f64) => f64 {
    const pi = 3.141592653589793f32;
    const radius = (/ diameter 2);
    const circumference = (* (* 2 pi) radius);
    return circumference;
}
